features/375-static-cdn/design.md

```diff
@@ -0,0 +1,138 @@
+- Developers should be able to upload static content to be hosted at specific path.
+- Hosted static content should be served & cached with CDN if possible.
+- SPA style routing should be supported.
+
+## Design
+
+### Static Deployment
+
+A new type of deployment, `static`, would be supported. This deployment type
+supports configurating a deployment context (`context`) and fallback absolute
+path (`fallback`).
+
+When `static` deployment type is used:
+- The deployment context would be persisted to object storage.
+- A request with matching path would be routing using static deployment routing
+  algorithm (described below).
+- If a file is matched, the gateway would proxy the request to the backing
+  object storage file.
+- If no file is matched, a generic 404 page would be returned.
+- The response would have appropiate cache control headers, as configured by
+  developer.
+
+### Static Deployment Routing
+
+When a request matched a `static` deployment, only its path would be considered
+during routing. The request would be routed as followed:
+
+1. Let `U` be the request URI, let `P` be the routing path without trailing slashes.
+2. If file `P` exists, return `P`.
+3. If file `Q` = `P/index.html` exists, return `Q`.
+4. If `fallback` is configured, set `U` = `U` with path `fallback` and route again.
+5. Otherwise, no file is matched.
+
+This procedure is similar to nginx directives:
+```
+index index.html
+try_files $uri <fallback>
+```
+
+Sometimes routing procedure is restarted (e.g. in step 4). The routing procedure
+can be performed at most 5 times to prevent infinite loop. In this case, `500
+Internal Server Error` response would be produced.
+
+**Example**:
+
+Suppose developer configured a `static` deployment as follow:
+```yaml
+  - name: assets
+    type: static
+    path: /
+    context: ./static
+    fallback: /index.html
+```
+
+- **https://example.com/main.js**:  
+    Route to `assets` and serve file at `./static/main.js`. (step 2)  
+- **https://example.com/**:  
+    Route to `assets` and serve file at `./static`. (step 2)  
+    Not a file, so instead serve file at `./static/index.html`. (step 3)  
+- **https://example.com/index.html**:  
+    Route to `assets` and serve file at `./static/index.html`. (step 2)  
+- **https://example.com/login**:  
+    Route to `assets` and serve file at `./static/login`. (step 2)  
+    File is not found, so instead serve file at `./static/login/index.html`. (step 3)  
+    File is not found, so instead rewrite URL with fallback path
+    `https://example.com/index.html` and route again. (step 4)  
+    Route to `assets` and serve file at `./static/index.html`. (step 1)
+
+### Caching
+
+Developer can configure cache duration per `static` deployment. The gateway
+would attach appropiate headers to response according to configuration.
+
+By default, the cache duration is 1 hour. Developers can disable caching by
+setting it explicitly to zero, but the maximum cache duration is 7 days.
```

Comment:
I think the 1 hour default cache duration may surprise developer that, when they start using custom domain, they may not aware that there is CDN in between and found their site won't update after deploy. 

1. I think maybe we can invalidate the cache after deployment. Since we don't have idea about the user content, so we can only invalidate all. It is not good for performance, cdn will need to fetch all the content again.
2. @louischan-oursky suggest that maybe we can change expires field to mandatory, I also think it is good too. So when user fill in this number they will know that there is CDN in the setup.

Thought?

Also, I think we cannot avoid we need to have a separate command to allow developer invalidate the cache, may the simplest version is just allow invalidate all?

Reply:
> they may not aware that there is CDN in between and found their site won't update after deploy.

I don't think it is surprising: this behavior can be understood quickly by inspecting the response headers, and many major static hosting sites require some times for the static content to update after deployment (e.g. GitHub pages).

Also, making `expires` field mandatory could also confuse the developer: if developers changed the expiry time (e.g. from 1 hour to zero), it would not be reflected immediately, and allow configuring expiry time does not implies the presence of CDN.

In my opinion, I think the current behavior as specified is acceptable, and I think it is a common expectation that it takes some time to update cached static content after deployment.

Reply:
Maybe we can have copywriting in the successful message (maybe after deploy or configure custom domain), and said there is 1 hour cache for static asset by default?

But do you think we should have a way for user to invalidate cache? In case the developer find some serious mistake after deployment and want to update the cdn immediately.

Reply:
> Maybe we can have copywriting in the successful message (maybe after deploy or configure custom domain), and said there is 1 hour cache for static asset by default?

I think this is good enough, something like 'Your deployment is successful. It may take some time to update cached CDN content'.

> But do you think we should have a way for user to invalidate cache? In case the developer find some serious mistake after deployment and want to update the cdn immediately.

I'd say this is not a must-have feature that we have to implement now. We can revisit it later when enough users request this.

Reply:
I still think invalidation is an important feature. If the developer make a mistake, they can help themselves instead of looking for support from us.

Reply:
> I still think invalidation is an important feature. If the developer make a mistake, they can help themselves instead of looking for support from us.

As mentioned before, many major static website hosts with CDN (e.g. Github Pages, now.sh, surge.sh) does not offer this option, so I don't think it is important.

For example, If I made a mistake when deploying on GitHub Pages, I would expect it takes up to an hour for the CDN cache to update.

On the other hand, some CDN (now.sh, surge.sh) supports cache invalidation on deployment, we can research on how to support this in a later iteration?

Also, for built-in CDN of cloud providers, cache invalidation seems to be rather restricted: GCP Cloud CDN limits at most one invalidation per minute, and Azure CDN have low concurrent invalidation request limit (~50).

Reply:
For serious use case some mistakes can't wait an unknown length of time. Be more specific / have a way to invalidate is better.

Reply:
Can you give examples of what use cases and mistakes we should be considering in the discussion? Given that there exists at least one major production-grade static page host (GitHub Page) that does not support invalidation, I don't think it is essential to support it now.

Reply:
Github Page is definitely not very production considering its a free services and most usage are from personal webpage to random note.

An example could be, someone posted a mobile phone no, or a wrong cost on the website, and need to be invalidated immediately, or each minute of delay would result is monetory loss

Reply:
> Github Page is definitely not very production considering its a free services and most usage are from personal webpage to random note.

I disagree on this. Free services can be production grade, and many large projects uses GitHub page as main site host (e.g. https://angular.io/, https://facebook.github.io/react-native/, https://getbootstrap.com/).

> An example could be, someone posted a mobile phone no, or a wrong cost on the website, and need to be invalidated immediately, or each minute of delay would result is monetory loss

Cache invalidation is also not an immediate operation; for CloudFront, it takes 10 - 15 minutes to complete as mentioned in documentation. Also, we cannot invalidate client-side/intermediate proxy cache. If developer is concerned about this, they can disable caching on the page or use a short cache duration (e.g. 10 mins).

Reply:
> I disagree on this. Free services can be production-grade, and many large projects use GitHub page as main site host (e.g. https://angular.io/, https://facebook.github.io/react-native/, https://getbootstrap.com/).

Well but all those are not very mission-critical to the business, mostly simple landing page. e.g. I haven't encountered any e-commerce or serious stuff running on it. That's what I meant

> Cache invalidation is also not an immediate operation; for CloudFront, it takes 10 - 15 minutes to complete as mentioned in documentation. Also, we cannot invalidate client-side/intermediate proxy cache. If developer is concerned about this, they can disable caching on the page or use a short cache duration (e.g. 10 mins).

That's why I said "For serious use case some mistakes can't wait an unknown length of time. Be more specific / have a way to invalidate is better."; Even without cache invalidation, nothing is "immediate"...

The unknown is a configuration with a default long values, someone will bombard us for support someday for this, I think we should either:

1. Make the default value short, maybe 10 mins as you proposed is good enough
2. Netlify's approach to this is it adds a fingerprint to all files, but I don't think we can replicate that...?
3. Make the value explicit
4. Enable cache invalidation feature

Reply:
> Well but all those are not very mission-critical to the business, mostly simple landing page. e.g. I haven't encountered any e-commerce or serious stuff running on it. That's what I meant

I think we have different definitions on what is 'production-grade'.

I'll make the cache expiry duration explicit for now.